qos.go

```diff
@@ -16,12 +16,188 @@
+	namedUUID, err := newRowUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	row := make(OVNRow)
+	row["direction"] = direction
+	if priority != nil {
+		row["priority"] = *priority
+	}
+	row["match"] = match
+
+	if action != nil {
+		oMap, err := libovsdb.NewOvsMap(action)
+		if err != nil {
+			return nil, err
+		}
+		row["action"] = oMap
+	}
+	if bandwidth != nil {
+		oMap, err := libovsdb.NewOvsMap(bandwidth)
+		if err != nil {
+			return nil, err
+		}
+		row["bandwidth"] = oMap
+	}
+	if external_ids != nil {
+		oMap, err := libovsdb.NewOvsMap(external_ids)
+		if err != nil {
+			return nil, err
+		}
+		row["external_ids"] = oMap
+	}
+
+	insertOp := libovsdb.Operation{
+		Op:       opInsert,
+		Table:    tableQoS,
+		Row:      row,
+		UUIDName: namedUUID,
+	}
+
+	mutateUUID := []libovsdb.UUID{{namedUUID}}
+	mutateSet, err := libovsdb.NewOvsSet(mutateUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	mutation := libovsdb.NewMutation("qos_rules", opInsert, mutateSet)
+	condition := libovsdb.NewCondition("name", "==", ls)
+
+	mutateOp := libovsdb.Operation{
+		Op:        opMutate,
+		Table:     tableLogicalSwitch,
+		Mutations: []interface{}{mutation},
+		Where:     []interface{}{condition},
+	}
+	operations := []libovsdb.Operation{insertOp, mutateOp}
+
+	return &OvnCommand{operations, odbi, make([][]map[string]interface{}, len(operations))}, nil
+}
+
+func (odbi *ovnDBImp) delQoSImp(ls string, direction string, priority *int, match string) (*OvnCommand, error) {
+	row := make(OVNRow)
+	if len(direction) > 0 {
+		row["direction"] = direction
+	}
+	if priority != nil {
+		row["priority"] = *priority
+	}
+	if len(match) > 0 {
+		row["match"] = match
+	}
+
+	qosUUIDs := odbi.getRowUUIDs(tableQoS, row)
+	if len(qosUUIDs) == 0 {
+		return nil, ErrorNotFound
+	}
+
+	var ovsdbUUIDs []libovsdb.UUID
+	for _, uuid := range qosUUIDs {
+		ovsdbUUIDs = append(ovsdbUUIDs, libovsdb.UUID{uuid})
+	}
+	uuidSet, err := libovsdb.NewOvsSet(ovsdbUUIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	conditions := libovsdb.NewCondition("_uuid", "==", uuidSet)
+	deleteOp := libovsdb.Operation{
+		Op:    opDelete,
+		Table: tableQoS,
+		Where: []interface{}{conditions},
+	}
+
+	mutation := libovsdb.NewMutation("qos_rules", opDelete, uuidSet)
+	mucondition := libovsdb.NewCondition("name", "==", ls)
+	mutateOp := libovsdb.Operation{
+		Op:        opMutate,
+		Table:     tableLogicalSwitch,
+		Mutations: []interface{}{mutation},
+		Where:     []interface{}{mucondition},
+	}
+	operations := []libovsdb.Operation{deleteOp, mutateOp}
+	return &OvnCommand{operations, odbi, make([][]map[string]interface{}, len(operations))}, nil
+}
+
+func (odbi *ovnDBImp) listQoSImp(ls string) ([]*QoS, error) {
+	var listQoS []*QoS
+
+	odbi.cachemutex.RLock()
+	defer odbi.cachemutex.RUnlock()
+
+	cacheLogicalSwitch, ok := odbi.cache[tableLogicalSwitch]
+	if !ok {
+		return nil, ErrorNotFound
+	}
+
+	for _, drows := range cacheLogicalSwitch {
```

Comment:
is it possible to have common function for that, so many duplicated code now. 

interface := drows.Fields["xxxx"]
interface.(type)
switch.....

not sure it works.

Reply:
i'm try to think about this, but now i don't know how to do that.